src/main/kotlin/org/rust/lang/core/macros/MacroGraphWalker.kt

```diff
@@ -0,0 +1,125 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.core.macros
+
+import com.intellij.lang.PsiBuilder
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.TextRange
+import org.rust.lang.core.macros.MGNodeData.*
+import org.rust.lang.core.parser.createRustPsiBuilder
+import java.util.*
+import kotlin.math.min
+
+/**
+ * Walks the [graph] along with macro [callBody] and determines possible [FragmentKind]s for the given [caretOffset]
+ */
+class MacroGraphWalker(
+    project: Project,
+    private val graph: MacroGraph,
+    private val callBody: String,
+    private val caretOffset: Int
+) {
+    private enum class Status { Active, Dead, Finished }
+
+    data class FragmentDescriptor(
+        val fragmentText: String,
+        val caretOffsetInFragment: Int,
+        val kind: FragmentKind
+    )
+
+    data class State(
+        val position: MacroGraphNode,
+        val marker: PsiBuilder.Marker?,
+        val descriptor: FragmentDescriptor?
+    )
+
+    private fun rollbackToState(state: State) {
+        position = state.position
+        state.marker?.rollbackTo()
+        descriptor = state.descriptor
+    }
+
+    private val builder = project.createRustPsiBuilder(callBody)
```

Comment:
Run `builder.eof()` before the first usage of the builder. Will be needed after rebasing to the master